### PR TITLE
[FLANE-132] Open ed x catalogue email service

### PR DESCRIPTION
**Pull request by UvgenGen**
[FLANE-132](https://youtrack.raccoongang.com/issue/FLANE-132) Open edX Catalogue Email Service

AC:
Users receive a report on email in XLS format
Email Subject: Catalogue Email Service
Admin can set users emails who can receive the report (in django admin ) + the Active checkbox
Weekly auto reports (for example on Monday)
On Sysadmin Dashboard:
Sysadmin can send the report manually from the Sysadmin Dashboard>Courses
Sysadmin can export the report from LMS
The report should include the next data:

1. Course Run ID (the last 7 digits of the URL)
2. Course URL
3. Course name
4. Enrollment end date
5. Course end date

### CODE REVIEW

File: pavelib/servers.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def celery(options):
     """
     Runs Celery workers.
     """
-    settings = getattr(options, 'settings', 'dev_with_worker')
+    settings = getattr(options, 'settings', 'devstack_with_worker')
```
Comment (cmltaWt0) [inline pavelib/servers.py]: Remove this.

File: lms/djangoapps/dashboard/admin.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from django.contrib import admin
+
+from dashboard.models import EmailsAdressMailing
```
Comment (OPersian) [inline lms/djangoapps/dashboard/admin.py]: @UvgenGen `EmailsAdressMailing` -> `EmailsAddressMailing` 
and in other places with `address` in vars' namings (e.g. `email_adress`)

File: lms/djangoapps/dashboard/sysadmin.py [hunk 3]
```diff
@@ -498,6 +499,16 @@ def post(self, request):
...
                         _('Deleted'), course.location.to_deprecated_string(), course.id.to_deprecated_string(), course.display_name)
 
+        elif action == 'get_courses_xls':
+            xls_file = get_courses_xls.delay()
+            response = HttpResponse(xls_file.result.getvalue(), content_type='application/vnd.ms-excel')
+            response['Content-Disposition'] = "attachment; filename=Courses_{0}.xls".format(request.META['SERVER_NAME'])
+            return response
+        
+        elif action == 'send_courses_xls':
+            email_adress = request.POST.get('email', '')
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/sysadmin.py]: Add validation for empty email (empty string).

File: lms/djangoapps/dashboard/tasks.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+"""
+This file contains celery tasks sysdashboard
+"""
+import logging
+import time
+import StringIO
+import xlwt
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Pep8 imports styling.
Please change:
https://www.python.org/dev/peps/pep-0008/#imports

Should be
```python
import logging
import time
import StringIO

import xlwt
from celery import task
from django.core.cache import cache
from django.conf import settings
from django.core.mail import EmailMessage, get_connection
from django.utils.translation import ugettext as _

from student.roles import CourseStaffRole, CourseInstructorRole
from xmodule.modulestore.django import modulestore

from .models import EmailsAddressMailing
```

File: lms/djangoapps/dashboard/tasks.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+from xmodule.modulestore.django import modulestore
+from .models import EmailsAddressMailing
+
+
+@task(name='mass_sending_report')
+def mass_sending_report():
+    connection = get_connection()
+    xls_file = get_courses_xls()
+    email_list = []
+    for email in EmailsAddressMailing.objects.all():
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: We can filter `active` emails here.

File: lms/djangoapps/dashboard/tasks.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        if email.active:
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: See comment above.

File: lms/djangoapps/dashboard/tasks.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            email_message = EmailMessage('Catalogue Email Service', 'Hi all,\nThe list of courses is in the attachment.', settings.EMAIL_FROM, [email.email],)
+            email_message.attach('Courses_report.xls', xls_file.getvalue(), 'application/vnd.ms-excel')
+            email_list.append(email_message)
+    connection.send_messages(tuple(email_list))
+    
+@task()
+def send_report_email(request, email_adress):
+    xls_file = get_courses_xls()
+    email_message = EmailMessage('Catalogue Email Service', 'Hi all,\nThe list of courses is in the attachment.', settings.EMAIL_FROM, [email_adress],)
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Typo - email_adress => email_address

File: lms/djangoapps/dashboard/tasks.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+@task()
+def get_courses_xls():
+    """
+    Getting course data and return xls file
+    _____________________________________________________________________________________
+    | Course URL | Course Run ID  | Course name | Enrollment end date | Course end date |
+    """
+    data = []
+    roles = [CourseInstructorRole, CourseStaffRole, ]
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Why do we need this list?

File: lms/djangoapps/dashboard/tasks.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    def_ms = modulestore()
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: What does `def_ms` mean? Defined modulestore? Please describe or change the name to something understandable.

File: lms/djangoapps/dashboard/tasks.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    for course in def_ms.get_courses():
+        if course.enrollment_end:
+            enroll_end_date = course.enrollment_end
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Please do it in one string:
```python
enroll_end_date = course.enrollment_end.strftime("%m/%d/%Y") if course.enrollment_end else ''
```

File: lms/djangoapps/dashboard/tasks.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            enroll_end_date = enroll_end_date.strftime("%m/%d/%Y")
+        else:
+            enroll_end_date = ''
+        if course.end:
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: The same as above.

File: lms/djangoapps/dashboard/tasks.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            end_date = course.end
+            end_date = end_date.strftime("%m/%d/%Y")
+        else:
+            end_date = ''
+        datum = [settings.LMS_ROOT_URL + "/courses/" + str(course.id), course.id.run, course.display_name,
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Please try this format instead:
```python
course_url = '{}{}'.format(
   settings.LMS_ROOT_URL,
   reverse('course_root', kwargs={'course_id': course.id})
)
```

File: lms/djangoapps/dashboard/tasks.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+            enroll_end_date = enroll_end_date.strftime("%m/%d/%Y")
+        else:
+            enroll_end_date = ''
+        if course.end:
+            end_date = course.end
+            end_date = end_date.strftime("%m/%d/%Y")
+        else:
+            end_date = ''
+        datum = [settings.LMS_ROOT_URL + "/courses/" + str(course.id), course.id.run, course.display_name, 
+                enroll_end_date, end_date, ]
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Remove trailing `,` and space.

File: lms/djangoapps/dashboard/tasks.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    wb = xlwt.Workbook()
+    ws = wb.add_sheet('Courses')
+
+    style = xlwt.XFStyle()
+    pattern = xlwt.Pattern()
+    pattern.pattern = xlwt.Pattern.SOLID_PATTERN
+    pattern.pattern_fore_colour = xlwt.Style.colour_map['gray25']
+    style.pattern = pattern
+
+
```
Comment (cmltaWt0) [inline lms/djangoapps/dashboard/tasks.py]: Too many blank lines.

File: lms/envs/aws.py [hunk 15]
```diff
@@ -131,7 +131,18 @@
...
+    })
+
+if FEATURES["ENABLE_SYSADMIN_DASHBOARD"]:    
+    CELERYBEAT_SCHEDULE.update({
+        # send an email with a report on the courses to the mail 
+        # specified in the model EmailsAddressMailing 
+        # every Monday morning at 11:30 A.M
+        "mass-sending-report": {
+            "task": "mass_sending_report",
+            "schedule": crontab(hour=11, minute=30, day_of_week=1),
```
Comment (cmltaWt0) [inline lms/envs/aws.py]: Is this an approved schedule?

**Pull request merged**